### PR TITLE
Typed KLEE

### PR DESCRIPTION
**Pull request by getN1ght**
## Checklist:
- [x] The PR addresses a single issue.  If it can be divided into multiple independent PRs, please do so.
- [x] The PR is divided into a logical sequence of commits OR a single commit is sufficient.
- [x] There are no unnecessary commits (e.g. commits fixing issues in a previous commit in the same PR).
- [x] Each commit has a meaningful message documenting what it does.
- [x] All messages added to the codebase, all comments, as well as commit messages are spellchecked.
- [x] The code is commented OR not applicable/necessary.
- [x] The patch is formatted via clang-format OR not applicable (if explicitly overridden leave unchecked and explain).
- [x] There are test cases for the code you added or modified OR no such test cases are required.

### CODE REVIEW

File: lib/Core/Executor.cpp [hunk 1]
```diff
@@ -139,6 +142,10 @@ cl::opt<bool>
                       cl::desc("Enable lazy instantiation (default=true)"),
                       cl::cat(ExecCat));
 
+cl::opt<bool> StrictAliasingRule(
+    "strict-aliasing",
+    llvm::cl::desc("Turn's on rules based on strict aliasing rule"),
```
Comment (misonijnik) [inline lib/Core/Executor.cpp:147]: Add information about the default value

File: lib/Core/Executor.cpp [hunk 2]
```diff
@@ -675,20 +689,19 @@ void Executor::initializeGlobalObject(ExecutionState &state, ObjectState *os,
...
-  ObjectState *os = bindObjectInState(state, mo, false);
+  ObjectState *os = bindObjectInState(state, mo, type, false);
   for(unsigned i = 0; i < size; i++)
     os->write8(i, ((uint8_t*)addr)[i]);
   if(isReadOnly)
     os->setReadOnly(true);  
   return mo;
 }
 
-
```
Comment (misonijnik) [inline lib/Core/Executor.cpp:691]: ?

File: lib/Core/Executor.cpp [hunk 3]
```diff
@@ -737,8 +751,13 @@ void Executor::allocateGlobalObjects(ExecutionState &state) {
 
 #ifndef WINDOWS
   int *errno_addr = getErrnoLocation(state);
+  llvm::Type *pointerErrnoAddr = llvm::PointerType::get(
+      llvm::IntegerType::get(m->getContext(), sizeof(*errno_addr) * 8),
```
Comment (misonijnik) [inline lib/Core/Executor.cpp:755]: Use named constants instead of literals

File: lib/Core/Executor.cpp [hunk 4]
```diff
@@ -4348,8 +4403,13 @@ void Executor::callExternalFunction(ExecutionState &state,
       addConstraint(state, EqExpr::create(ce, *ai));
       ObjectPair op;
       // Checking to see if the argument is a pointer to something
+      llvm::Type *argumentType = nullptr;
+      if (ati != function->arg_end()) {
```
Comment (misonijnik) [inline lib/Core/Executor.cpp:4407]: Add a comment

File: lib/Core/Executor.cpp [hunk 5]
```diff
@@ -4684,9 +4756,9 @@ void Executor::resolveExact(ExecutionState &state,
 
 void Executor::executeMemoryOperation(ExecutionState &state,
                                       MemoryOperation operation,
-                                      ref<Expr> address,
-                                      ref<Expr> value /* undef if read */,
-                                      KInstruction *target /* undef if write */) {
+                                      KType *targetType, ref<Expr> address,
+                                      ref<Expr> value /* def if write*/,
```
Comment (misonijnik) [inline lib/Core/Executor.cpp:4760]: Leave it the same

File: lib/Core/Executor.cpp [hunk 6]
```diff
@@ -4827,53 +4918,88 @@ void Executor::executeMemoryOperation(ExecutionState &state,
       ExecutionState *bound_inner = branches_inner.first;
       ExecutionState *unbound_inner = branches_inner.second;
       if (bound_inner) {
+        ObjectState *wos = bound_inner->addressSpace.getWriteable(mo, os);
```
Comment (misonijnik) [inline lib/Core/Executor.cpp:4921]: Maybe it's better to call getWriteable only when the type in the ObjectState changes? At least write a comment

File: lib/Core/Executor.cpp [hunk 7]
```diff
@@ -5058,10 +5194,10 @@ ExecutionState* Executor::formState(Function *f,
     arguments.push_back(ConstantExpr::alloc(argc, Expr::Int32));
     if (++ai!=ae) {
       Instruction *first = &*(f->begin()->begin());
-      argvMO =
-          memory->allocate((argc + 1 + envc + 1 + 1) * NumPtrBytes,
-                           /*isLocal=*/false, /*isGlobal=*/true,
-                           /*allocSite=*/first, /*alignment=*/8);
+      argvMO = memory->allocate((argc + 1 + envc + 1 + 1) * NumPtrBytes,
```
Comment (misonijnik) [inline lib/Core/Executor.cpp:5197]: Revert changes

**Review by misonijnik — CHANGES_REQUESTED** [review]

File: lib/Core/Memory.cpp [hunk 8]
```diff
@@ -127,6 +130,10 @@ ObjectState::ObjectState(const ObjectState &os)
   memcpy(concreteStore, os.concreteStore, size*sizeof(*concreteStore));
 }
 
+ObjectState::ObjectState(const ObjectState &os, KType *dt) : ObjectState(os) {
+  dynamicType = dt;
```
Comment (ladisgin) [inline lib/Core/Memory.cpp:134]: move to initialization list

**Review by ladisgin — CHANGES_REQUESTED** [review]

File: lib/Core/Executor.cpp [hunk 9]
```diff
@@ -139,6 +142,21 @@ cl::opt<bool>
                       cl::desc("Enable lazy instantiation (default=true)"),
                       cl::cat(ExecCat));
 
+cl::opt<TypeSystemKind>
+    TypeSystem("type-system",
+               cl::desc("Use information about type system from specified "
+                        "language (default=llvm)"),
+               cl::values(clEnumValN(TypeSystemKind::LLVM, "LLVM",
```
[the remark this answers is not in the document]
Comment (getN1ght) [reply lib/Core/Executor.cpp:149]: As KLEE hardly support any other language than C/C++, we do not need it yet

File: lib/Core/Memory.cpp (hunk 8 above)
Comment (getN1ght) [reply lib/Core/Memory.cpp:134]: ?

**Review by misonijnik — APPROVED** [review]

**Review by ladisgin — APPROVED** [review]

**Pull request merged**